source/simulator/assembly.cc

```diff
@@ -2248,34 +2307,69 @@ namespace aspect
+      }
 
-    assemblers->local_assemble_advection_system
-    .connect (std_cxx11::bind(&aspect::Assemblers::CompleteEquations<dim>::local_assemble_advection_system,
-                              std_cxx11::cref (*complete_equation_assembler),
-                              // discard cell,
-                              std_cxx11::_2,
-                              std_cxx11::_3,
-                              std_cxx11::_4,
-                              std_cxx11::_5));
-
-    assemblers->compute_advection_system_residual
-    .connect (std_cxx11::bind(&aspect::Assemblers::CompleteEquations<dim>::compute_advection_system_residual,
-                              std_cxx11::cref (*complete_equation_assembler),
-                              // discard cell,
-                              std_cxx11::_2,
-                              std_cxx11::_3));
+    if (parameters.include_melt_transport)
+      {
+        assemblers->local_assemble_advection_system
+        .connect (std_cxx11::bind(&aspect::Assemblers::MeltEquations<dim>::local_assemble_advection_system_melt,
+                                  std_cxx11::cref (*melt_equation_assembler),
+                                  // discard cell,
+                                  std_cxx11::_2,
+                                  std_cxx11::_3,
+                                  std_cxx11::_4,
+                                  std_cxx11::_5));
+
+        assemblers->compute_advection_system_residual
+        .connect (std_cxx11::bind(&aspect::Assemblers::MeltEquations<dim>::compute_advection_system_residual_melt,
+                                  std_cxx11::cref (*melt_equation_assembler),
+                                  // discard cell,
+                                  std_cxx11::_2,
+                                  std_cxx11::_3));
+      }
+    else
+      {
+        assemblers->local_assemble_advection_system
+        .connect (std_cxx11::bind(&aspect::Assemblers::CompleteEquations<dim>::local_assemble_advection_system,
+                                  std_cxx11::cref (*complete_equation_assembler),
+                                  // discard cell,
+                                  std_cxx11::_2,
+                                  std_cxx11::_3,
+                                  std_cxx11::_4,
+                                  std_cxx11::_5));
+
+        assemblers->compute_advection_system_residual
+        .connect (std_cxx11::bind(&aspect::Assemblers::CompleteEquations<dim>::compute_advection_system_residual,
+                                  std_cxx11::cref (*complete_equation_assembler),
+                                  // discard cell,
+                                  std_cxx11::_2,
+                                  std_cxx11::_3));
+      }
 
     if (parameters.use_discontinuous_temperature_discretization ||
         parameters.use_discontinuous_composition_discretization)
       {
+        AssertThrow(!parameters.include_melt_transport, ExcNotImplemented());
+
```

Comment:
This line really does just a sanity check that the set of parameters that have been handed through the input file make sense. I think this check would better fit into `Parameters::parse_parameters()` in `parameters.cc`.